eRFC: if- and while-let-chains, take 2

Centril opened the pull request:
### [🖼️ Rendered](https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md)

### 📝 Summary

Extends `if let` and `while let`-expressions with chaining, allowing you to combine multiple `let`s and `bool`-typed conditions together naturally. After implementing this RFC, you'll be able to write, among other things:

```rust
fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> ParamEnv<'tcx> {
    if let Some(Def::Existential(_)) = tcx.describe_def(def_id)
        && let Some(node_id) = tcx.hir.as_local_node_id(def_id)
        && let hir::map::NodeItem(item) = tcx.hir.get(node_id)
        && let hir::ItemExistential(ref exist_ty) = item.node
        && let Some(parent) = exist_ty.impl_trait_fn
    {
        return param_env(tcx, parent);
    }

    ...
}
```

The main aim of this RFC is to decide that this is a problem worth solving as well as discussing a few available options. **Most importantly, we want to make `let PAT = EXPR && ..` a possible option for Rust 2018.**

### 💖 Thanks

To everyone who participated in RFC #2260 and to the survey participants.
To @scottmcm for collaborating with me on the original RFC.
To @aturon and @nikomatsakis for taking the time to discuss this with me.
To @SergioBenitez for checking the feasibility of this in Rust 2018.
To @oli-obk for providing me with the useful example in the summary ;)
To @joshtriplett for the consultation on the style.
To @kennytm for improving clarity on the Rust 2018 migration technical changes.

Ixrec commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2000 @@
+`let PAT = EXPR && ..` as a user is a bad thing to do.
+
+Finally, some may argue that this is *"a lot of ad-hoc syntax to deprecate when
+the proper solution solving all the listed problems is implemented"*.
```
What are the phrases "ad-hoc syntax" and "the proper solution" trying to refer to in this sentence?

Everything else in the RFC makes perfect sense and seems compelling to me.

SergioBenitez commented:
I am very _for_ this change. This syntax feels natural and intuitive, and it is something I've pined for.

My remarks are:
1. While I have often wanted to reach for `if let` chains, I've never instinctively wanted `while let` chains. As such, personally, I do not see `while let` chaining as a necessary addition.
2. With somewhat less frequency than `&&` chaining, I have also reached for `||` chaining in `if let` clauses. In particular, in combination with `&&`:
    ```rust
    if let A(x) = foo() || (let B(x) == bar() && x.is_this()) {
        // use x
    }
    ```
    Again, this construction and syntax feel natural to me, and I would consider it `if let` chaining _without_ `||` to be inconsistent with the rest of the language. That being said, I'd rather have `&&` chaining alone than nothing at all.

Centril commented:
@SergioBenitez 

> I've never instinctively wanted while let chains. As such, personally, I do not see `while let` chaining as a necessary addition.

Thank you for the data point and the remarks.

I think the primary motivation for also changing `while let` is consistency with `if let` so as to reduce surprises for users. Uniformity is our friend in making teaching and learning easier :)

That said, if we consider iterating over multiple sources that produce elements in a zip-like fashion (but not strictly `.zip(..)` for iterators), then I think chaining `while let` can be useful. It can also be useful to chain `while let` with a side-condition as in `while let Foo(bar) && bar.is_special() { .. }`.

> Again, this construction and syntax feel natural to me, and I would consider it `if let` chaining without `||` to be inconsistent with the rest of the language. That being said, I'd rather have `&&` chaining alone than nothing at all.

I'm personally undecided on adding `||` here. What this RFC *does* do is to make `||` a possible addition in Rust 2018, if we so choose. For now, due to the time constraints of shipping the edition, I'd like to punt on adding `||` to the language to a future RFC and discussions.

kennytm commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+## An expected feature
+[an_expected_feature]: #an-expected-feature
+
+As demonstrated in [Appendix B], the syntax proposed in this RFC is already
```
This link is broken on GitHub at least.

Centril replied:
Also fixed :)

kennytm commented on text/0000-if-let-chains.md:
````diff
@@ -0,0 +1,2003 @@
+```rust
+if independent_condition,
+   let Alan(x)   = turing(),
+   let Alonzo(y) = church(x),
````
These `,`s should be `&&`s.

Centril replied:
Fixed :) Thanks!

oli-obk commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+
+# Appendix A - Style considerations
+
+How should the features introduced in this RFC be formatted?
```
6.

```rust
if let foo = bar &&
if let x = y &&
z <= 42
{
}
```

No indent at all, just a list of conditions

Centril replied:
Wouldn't this be the following instead?

```rust
if let foo = bar &&
if let x = y &&
if z <= 42 {
    ...
}
```

oli-obk replied:
Well since I agree now that the `if`-spammy syntax is not good:

```rust
if let foo = bar &&
let x = y &&
z <= 42 {
    ...
}
```

Centril replied:
Updated the style discussion with this version now :)

oli-obk commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+1. > `&&` as separator would require boolean expressions.
+
+2. > `&&` is for boolean expressions, and won't work right in generic usage.
+   > `let` would have to return a boolean which is weird and probably a breaking change.
```
Again something that would become obvious if `&& if let` were used instead of `&& let`

```rust
let x: bool = if let a...
```

Already exists today and needs a bool expression in the body and an else branch

Centril replied:
Associated discussion: https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md#expectation-that-let-pat--expr--bool

TL;DR: this expectation can be readily ruled out with (good) error messages and `if let` already induces this problem by making users think that it is the composition of `if` + `let PAT = EXPR`. I don't think it will be that big of a problem in practice.

Centril commented:
@aturon has asked me to clarify what exactly is being proposed to be accepted with this RFC. So here goes...

The precedent set by closing RFCs #2443, #2441, and #2429 is that *we don't accept RFCs reserving syntax without accepting a feature along with it*. Therefore, to be consistent with this policy, and for the purposes of making `if let PAT = EXPR && ...` a possible syntax in Rust 2018, *we need also to accept a feature along with the reservation of this syntax in Rust 2018.*

So the idea with this RFC is that we accept the full feature now (the *idea* of chaining...), but leave *finalizing the syntax itself* unresolved-ish as we did with #2071 (`existential type Foo: Bar;`). This will leave us room both to make the changes necessary in terms of the lint in Rust 2015 as well as the hard error in Rust 2018.

After the Rust 2018 has shipped, and when we have time, we can then implement the proposed syntax and experiment with it on nightly. *The syntax will still have to be finalized by some other decision*, such as with another RFC or on the tracking issue.

oli-obk commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+   > circuiting the evaluation.
+
+## Appendix B.7
+[Appendix B.7]: #appendix-b7
```
What's this appendix about?

Centril replied:
Added a note. Thanks for spotting it!

Centril commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+evaluates to `do_stuff_with(x)`. If no patterns match, the expression instead
+evaluates to `do_other_stuff()`.
+
+This RFC does not propose such a facility, but does not for close such
```
Indeed it should and [does](https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md#rollout-plan-and-transitioning-to-rust-2018). :)

Centril commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2003 @@
+warning may give the impression that the irrefutable pattern is refutable,
+or cast doubt on it making semantics possibly harder to grasp quickly.
+
+This is a tricky question, which we leave open for consideration during
```
> Clippy needs this if we want to be able to remove the `if_chain` macro.

I'm basically of the mind that we should do it...

> But I think this is a separate feature that can be added later with its own RFC, so we shouldn't block this one on this question

.. and I super agree that we should wait with the decision.

On the question of differentiating between creation and pattern matching, I don't think there should be such a divide. In fact, `let Ok(x) | Err(x) = expr;` is already an RFC accepted feature (RFC #2175) and we can already do irrefutable pattern matching in `let` statements, so there isn't such a divide today.

I don't think mixing in irrefutable matches will be too common, and so we shouldn't optimize for that uncommon case. Where it does occur, `let ident = expr` is already, in my view, differentiated enough by binding an identifier or a struct pattern as opposed to an enum variant pattern.

petrochenkov commented:
My opinion hasn't changed since https://github.com/rust-lang/rfcs/pull/2260#issuecomment-353780537.
I still think this is a half-assed solution that will likely block a proper solution in the future.

We should aim for supporting a convenient non-exhaustive pattern-matching expressions in general (`EXPR op PAT: bool`) rather than special case `if` and `while`, because it solves much more issues than chaining alone.

>Furthermore, as evidenced in RFC 2260, making EXPR is PAT, which has other problems we've previously noted, an expression is also tricky due to the non-obvious scoping rules for bindings it entails. Mainly because of this, support for `EXPR is PAT` has been slow to develop.

That's just not true, for expressions in `if` and `while` the rules can be exactly the same as in this RFC.

If "slow to develop" refers to my promise to write an RFC for pattern-matching expressions (https://github.com/rust-lang/rfcs/issues/2411#issuecomment-383320967), then the reason is that I was busy with more urgent things and deferred this work until 2018 edition completion.
Sigh, logistically this RFC is perfectly timed, since I still can't write a counter-RFC right now or spend too much time arguing.

aturon commented:
@petrochenkov Totally understood -- did you see @Centril's [clarifying comment](https://github.com/rust-lang/rfcs/pull/2497#issuecomment-404734990)? My understanding is that this RFC is meant to commit us to (1) *some* solution to this problem and (2) carving out a bit of space in the grammar so that the solution *could* be as discussed in the RFC.

FWIW, I also have significant misgivings about solving this problem by essentially extending the special-cased treatment of `if let`, and would prefer something more compositional. **But** I am in favor of the commitments being proposed.

@Centril I wonder if this should be made formally an "eRFC" to signify more clearly the issues above?

petrochenkov commented:
@aturon 
Oh, if this is actually about reserving `if let PAT = EXPR1 && EXPR2` as `if (let PAT = EXPR1) && EXPR2` in the grammar, then I'm on board 😄 
I'm just worrying that the accepted version will become set in stone eventually.

kennytm commented:
[These operators](https://github.com/rust-lang-nursery/reference/blob/master/src/expressions.md#expression-precedence) all have lower precedence than `&&`:

* `||`
* `..`, `..=`
* `=`, `+=`, etc 
* `return`, `break`

Which means we also have the following ambiguity to consider:

```rust
    if let Range { start: _, end: false } = true..true && false {
        println!("(current behavior)");
    } else {
        panic!();
    }
```

```rust
    let result = loop {
        if let Some(0u32) = break true && false {
            panic!();
        }
        panic!();
    };
    assert_eq!(result, false);
```

And then `||` and `&&` can be interpreted prefix operators:

```rust
    const F: fn() -> bool = || true;
    if let Range { start: F, end } = F..|| false {
        println!("(current behavior)");
        assert!(!end());
    } else {
        panic!();
    }
```

```rust
    let t = &&true;
    if let Range { start: true, end } = t..&&false {
        println!("(current behavior)");
        assert_eq!(**end, false);
    } else {
        panic!();
    }
```

----

Please also ensure the parser *accepts* the following:

```rust
if a || b {}   // <-- no parenthesis, it is ok.
if a && b || c && d {}
// interpret as: `(a && b) || (c && d)`
```

and *rejects* the following before `if-let-or` is supported:

```rust
if let Some(a) = b && c || d {}
// interpret as: `((let Some(a) = b) && c) || d`
```

Centril commented:
@aturon eRFC it is ;)

@petrochenkov Yep, it's about making the proposed solution possible and recognizing that *some* solution (which could be `EXPR match PAT`) is necessary.

I would personally be fine with experimenting with `if EXPR match PAT && ..` and `if let PAT = EXPR && ..` *concurrently* in the nightly compiler and see which one comes out on top.

> That's just not true, for expressions in if and while the rules can be exactly the same as in this RFC, and outside of if/while these expressions can desugar into if expr { true } else { false } (at least this is how it's done in my implementation).

I see; my impression was that it was more complicated from the discussions at #2260. This is the "bindings only live until the end of a full expression" rule?

> If "slow to develop" refers to my promise to write an RFC for pattern-matching expressions
based on my implementation (#2411 (comment)),

Not at all :) It is referring to my perception of the support for `EXPR op PAT` within the community and the lang team (which could be wrong, but the survey is at least quite suggestive as to the community support).

> then the reason is that I was busy with more urgent things and deferred this work until 2018 edition completion.

I thank you for it ❤️

joshtriplett commented:
I can't say I'm a huge fan of the specific proposed syntax, but I'm inclined to take this step regardless, so that even if we don't end up going with this particular syntax we can detect potentially ambiguous uses of it. I'm honestly tempted to argue that either grouping should require parentheses.

Minor nit: for consistency, can you please always put `&&` at the start of a line, rather than at the end? In addition to matching the standard style/rustfmt, I also think that more clearly distinguishes the `&&` from the expression on the right-hand side of the `let`.

Also, in the example of how `while let` would translate to a `loop`, you have unbalanced `{`s at the ends of several lines.

joshtriplett commented on text/0000-if-let-chains.md:
```diff
@@ -0,0 +1,2000 @@
+evaluates to `do_stuff_with(x)`. If no patterns match, the expression instead
+evaluates to `do_other_stuff()`.
+
+This RFC does not propose such a facility, but does not for close such
```
*foreclose, or alternatively exclude

Centril commented:
Updates:
+ After discussion with @kennytm on Discord I've adjusted the Rust 2018 changes necessary to be more clear wrt. changes.
+ After discussion with @joshtriplett I've adjusted the style used in the RFC to conform to how `rustfmt` would deal with this. This is style 1. in the [Appendix A](https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md#appendix-a---style-considerations) which is provided for the style team's benefit.

kennytm commented on text/0000-if-let-chains.md:
````diff
@@ -574,24 +574,75 @@ There exists an ambiguity in this new grammar in how to parse:
+```rust
+if (let Range { start: _, end: _ } = true..true) && false { ... }
+
+if (let PAT = break true && false) { ... }
````
I think you mean `if (let PAT = break true) && false { ... }` here.

Centril replied:
Indeed; I think I did a copy error here - thanks for spotting it!

comex commented:
> I still think this is a half-assed solution that will likely block a proper solution in the future (for social rather than technical reasons).

> We should aim for supporting a convenient non-exhaustive pattern-matching expressions in general (`EXPR op PAT: bool`) rather than special case `if` and `while`, because it solves much more issues than chaining alone.

And I, on the other hand, still think the syntax ought to be`(let PAT = EXPR): bool`, making this RFC a half-assed solution that will *encourage* a proper solution in the future. ;)

petrochenkov commented:
@comex 
Pattern-matching expression must be as convenient as equality comparison
```rust
let y = x == Variant;
ley y = x op Variant;
```
, otherwise the problem is not solved.

Look at this:
```rust
let y = let Variant = x;
// or with parens, which may be required due to parsing ambiguities with `let`
let y = (let Variant = x);
```
This is painful to look at due to dissimilarity to equality - due to incorrect operand order, due to looking like a double assignment, due to an extra unnecessary token.

So you still have to generate boilerplate to hide this code fragment, like you have to do with `if let Variant = x { true } else { false }`.
- The most common approach right now - you need to implement `PartialEq` (usually with derive, but perhaps manually) to use equality instead of matching (`x == Variant`), even if it doesn't make sense for some variants (see e.g. `enum Token` in the compiler). Have you seen what these `derive(PartialEq)` expand into? Compiler has to process all that.
- Otherwise, you need to implement `fn is_variant() -> bool { (let Variant = x) }` boilerplate if `PartialEq` cannot be generated.

My goal for pattern matching expressions is to be pervasively used for comparisons with enum variants and elimination of `derive(PartialEq)`/`fn is_variant` boilerplate, so these expressions should look like comparisons and should be as convenient as comparisons.
So this sentence from the RFC
>Another group found that `is` could easily be confused for some sort of equality test (as in the operator `==`)

is not a drawback at all, pattern matching *is* a sort of equality test.

Centril commented:
@petrochenkov Before I address your other points, let's drill in to this one in particular:

> ```rust
> let y = let Variant = x;
> // or with parens, which may be required due to parsing ambiguities with `let`
> let y = (let Variant = x);
> ```

What parsing ambiguities do you foresee arising with `let`?

If we make `let PAT = EXPR` itself an expression, then the statement `let PAT = EXPR;` (note the semi) will be parsed as happens with `foo();`. If we instead consider `{ let PAT = EXPR }`, then it is not legal and so we can give it meaning.

petrochenkov commented:
@Centril 
`let x = y && z` first of all (previously discussed in https://github.com/rust-lang/rfcs/pull/2260#issuecomment-354573858 and following comments).
This is what this RFC future-proofs against, but only in `if` and `while`.

Nemo157 commented:
One variant that wasn't in the survey is

```rust
if EXPR is let PAT
    && EXPR is let PAT
    && EXPR
{
    ...
}
```

I find that avoids the main issue that `EXPR is PAT` doesn't seem like something that should introduce a binding, it also is consistent with a non-binding `EXPR is PAT` for when you just want to test without introducing a binding (this could deny any named bindings, forcing you to use `_` instead). It should also avoid syntax ambiguities as `PAT && ANYTHING` is not a valid `PAT`.

(Prior art for this could be [C# 6's declaration expressions](https://intellitect.com/declaration-expressions-in-c-6-0/)).

petrochenkov commented:
@Nemo157 
This is an interesting idea.
If `let` is optional syntactically and then enforced by a semantic check if `PAT` introduces bindings, then we'll be able to easily make it fully optional in the future if the extra `let` becomes too annoying (i.e. [Stroustrup's rule](https://thefeedbackloop.xyz/stroustrups-rule-and-layering-over-time/) will come into effect).

This general scheme  - `EXPR is PAT` for binding-less matching and `EXPR is PAT with some marker on PAT` for any matching would resolve by concerns about future-compatibility.

kennytm commented:
`EXPR is PAT` is still ambiguous because `is` [cannot be a keyword](https://doc.rust-lang.org/std/any/trait.Any.html#method.is).

```rust
#![feature(pattern_parentheses)]

use std::ops::RangeFrom;

fn is(a: RangeFrom<i32>) -> i32 {
    a.start
}

fn main() {
    if let (RangeFrom { start: 3 }) = 1.. {}
    let x = 1.. is (RangeFrom { start: 3 });
}
```

petrochenkov commented:
@kennytm 
This particular example doesn't seem to be ambiguous with current grammar - if `as` operator cannot be attached to some expression (e.g. a range `1..`), then `is` as an operator cannot be attached to it as well.
(Although it may be a parser limitation that `BEGIN.. as TYPE` doesn't work, but since we don't have a formal grammar...)
(And it doesn't matter anyway, there's big difference between common case ambiguities like `let x = y && z`and corner case ambiguities like this, `dyn` was also ambiguous in corner cases.)

kennytm commented:
@petrochenkov How is `as` relevant 😕? `as` is a keyword, and the RHS is a type, not a pattern.

kennytm commented:
@Centril What we still parse `if let x = y && z` as `if let x = (y && z)`, and you need to explicitly write `if (let x = y) && z`?

```rust
    if (let Some(Def::Existential(_)) = tcx.describe_def(def_id))
        && (let Some(node_id) = tcx.hir.as_local_node_id(def_id))
        && (let hir::map::NodeItem(item) = tcx.hir.get(node_id))
        && (let hir::ItemExistential(ref exist_ty) = item.node)
        && (let Some(parent) = exist_ty.impl_trait_fn) 
        // ^ the last one doesn't need brackets, but to be consistent...
    {
        return param_env(tcx, parent);
    }
```

This would be backward-compatible and doesn't need any edition lints.

Centril commented:
@petrochenkov Oh of course... We could raise the precedence of `&&` and `||` inside `let` as well... It doesn't seem like it would hit that many cases?

In any case, irrespective of whether we have to write `(let PAT = EXPR)` instead of `let PAT = EXPR`
let me agree with you that `let y = let Variant = x;` does not read well for the reasons you've stated.
However, we could regain effectively `is let` with:

```rust
macro_rules! is {
    ($(token:tt)+) => {
        if $(token)+ { true } else { false }
    }
}
```
in which case you could write:
```rust
let y = is!(let Variant = x);
```
which isn't *that* far off from `is let`. This trivial macro could be added to the standard library. You might object that we could just as well add `if_chain` to the standard library, but that is a much larger thing to standardize.

> My goal for pattern matching expressions is to be pervasively used for comparisons with `enum` variants and elimination of `derive(PartialEq)` / `fn is_variant` boilerplate, so these expressions should look like comparisons and should be as convenient as comparisons.

I agree with that goal in general.

> is not a drawback at all, pattern matching is a sort of equality test.

Yes it is a sort of equality test, but not all equalities are equal, and pretending like they are (with syntax which implies it) does not enhance understanding.

The operator `is` could very easily be misunderstood for `==`, which is user defined equality that might be *weaker* than pattern matching (e.g. if a field is not used in equality tests, c.f. `Vec<T: PartialEq>`, which does not care about the capacities being equal).

The operator could also be easily misunderstood for address-equality, which is considerably stronger than pattern matching is.

Furthermore, I have some additional misgivings about `is` specifically:
+ The token is lexically too close to `as`, `in` and `if`, making `is let` for example too close to `if let`.

Due to these concerns about binding and confusion with `as`, etc., I'm increasingly skeptical about `is` as an operator.

I also have some general concerns about `EXPR op PAT` (for example with `EXPR match PAT`):

+ It *is* inverted order since today patterns usually always occur to the LHS textually. And `match e { A(x) => ... }` does not count as a counter-example because you don't write it that way (there's always a newline per rustfmt: https://play.rust-lang.org/?gist=c2ecae9685202b1acc872083dd320724&version=stable&mode=debug&edition=2015)...

+ It doesn't immediately highlight introduction of bindings when reading left-to-right since bindings are introduced on RHS. Meanwhile, `if let PAT = EXPR` always introduces bindings on LHS which works better with LTR reading. Therefore, my conclusion is that `if let PAT = EXPR` reads better.

+ With `EXPR op PAT`, AIUI, you hit the problem of ambiguity with `if EXPR is Enum::Variant { ... } { ... }` which you then have to disambiguate as `if EXPR is (Enum::Variant { ... }) { ... }`. This seems like a bigger hit to ergonomics than having to write `is!(...)`.

+ It is another way to do the same thing. While Rust doesn't generally adhere to *"there should only be one way to do it"*, there should be some non-trivial reason why a new way to do it is introduced. In this case, I don't think it's that non-trivial.

@Nemo157 

> One variant that wasn't in the survey is

I only included the main options floated in #2260 :) I would pick `EXPR match PAT` instead of `EXPR is let PAT` because:
+ it is not two words.
+ it is lexically not close to `if let`
+ `match` indicates the binding nature well.
+ No new (contextual) keywords

@kennytm I'm skeptical that this should be required because it de-optimizes ergonomics to avoid breaking a corner case you should never write in any case and which simply doesn't occur in much (~any) code.
I'm doubly skeptical that we should not make the preparation for `if let PAT = EXPR && .. ` *now* because of the time constraints.

petrochenkov commented:
@Centril 
>It is inverted order since today patterns usually always occur to the LHS textually.

Patterns in "assignments"/"variable declarations" (`let x = 10;`) are on LHS and that's totally okay, they are not conditions and they are not should be consistent with conditions.
Patterns in `if let PAT = EXPR` being on the left side is a mistake, that I consciously don't want to follow. It's exactly my goal to be inconsistent with them, to fix the mistake.
The "guide-level explanation" is more or less "`let` means assignment/declaration, the pattern is on the left, the binding is long-living" and "`match`/`is` means condition, the pattern is on the right, the binding is "temporary"/short-living".

>It doesn't immediately highlight introduction of bindings when reading left-to-right since bindings are introduced on RHS. Meanwhile, if let PAT = EXPR always introduces bindings on LHS which works better with LTR reading.

This can be realistic issue that needs experimentation and can be fixed by `EXPR is let PAT`-like schemes for conditions with bindings if necessary.
The [converted version of rustfmt](https://github.com/petrochenkov/rustfmt/commits/isproto) looked pretty good to me, highlighting would improve things though.

>With EXPR op PAT, AIUI, you hit the problem of ambiguity with if EXPR is Enum::Variant { ... } { ... } which you then have to disambiguate as if EXPR is (Enum::Variant { ... }) { ... }.

This is also true and annoying, but also needs an estimation of how often this pattern can arise.
There was no such examples in rustfmt, but that's project-specific.
`EXPR` on the right has the same problem though, so this is something universally annoying rather than a  drawback of `EXPR op PAT`.

>This seems like a bigger hit to ergonomics than having to write is!(...).

Oh, come on.

Everything between "Yes it is a sort of equality test..." and "...I'm increasingly skeptical about is as an operator" looks like really far-fetched speculations to me, rooted rather in rhetorics than reality.

Centril commented:
>  Patterns in `if let PAT = EXPR` being on the left side is a mistake, that I consciously don't want to follow. It's exactly my goal to be inconsistent with them, to fix the mistake.

Well, we disagree that the pattern being on the LHS is a mistake there; I think it is good for readability.
And remember that [RFC 160 did have the option of going with `EXPR is PAT` or `EXPR match PAT`, but chose not to](https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md#rfc-160).

> The "guide-level explanation" is more or less "`let` means assignment/declaration, the pattern is on the left, the binding is long-living" and "`match`/`is` means condition, the pattern is on the right, the binding is "temporary"/short-living".

And what about:

```rust
match expr {
    A(x) => ...,
    B(y) => ...,
}
```

The pattern is on the *left* and "short lived".

> This can be realistic issue that needs experimentation and can be fixed by `EXPR is let PAT`-like schemes for conditions with bindings if necessary.
The converted version of rustfmt looked pretty good to me, highlighting would improve things though.

Aside: I've been writing this a few times and a lot of the time I'm writing `if let` instead of `is let` by mistake...

With `EXPR is let PAT`, the pattern is still on the RHS; at best, you can highlight `is let` differently than other keywords, but even with syntax highlighting I don't think it reads better. An example... Compare:

```rust
fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> ParamEnv<'tcx> {
    if tcx.describe_def(def_id) match Some(Def::Existential(_))
        && tcx.hir.as_local_node_id(def_id) match Some(node_id)
        && tcx.hir.get(node_id) match hir::map::NodeItem(item)
        && item.node match hir::ItemExistential(ref exist_ty)
        && exist_ty.impl_trait_fn match Some(parent)
    {
        return param_env(tcx, parent);
    }

    ...
}
```
with:

```rust
fn param_env<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId) -> ParamEnv<'tcx> {
    if let Some(Def::Existential(_)) = tcx.describe_def(def_id)
        && let Some(node_id) = tcx.hir.as_local_node_id(def_id)
        && let hir::map::NodeItem(item) = tcx.hir.get(node_id)
        && let hir::ItemExistential(ref exist_ty) = item.node
        && let Some(parent) = exist_ty.impl_trait_fn
    {
        return param_env(tcx, parent);
    }

    ...
}
```

In the latter case, the `let`s that introduce bindings align perfectly, while the `match`es in the above example don't.

> This is also true and annoying, but also needs an estimation of how often this pattern can arise.
There was no such examples in rustfmt, but that's project-specific.
`EXPR` on the right has the same problem though, so this is something universally annoying rather than a drawback of `EXPR op PAT`.

Enum tuple variants are probably more common by an order of magnitude, but I doesn't seem like a good idea to me to pessimize enum struct variants further. I also think it would be quite uncommon to pattern match with `if let PAT = Foo::Bar { ... } { .. }`. Usually when pattern matching, you are doing it against the return value of some function or some variable.

> Oh, come on.

I believe it is, because I don't think `let variable: bool = EXPR is PAT;` will be all that common. If you are doing some test, you usually want to act on that test, and so you have `if let PAT = EXPR`, which reads well to me and is ergonomic. If you disagree, please elaborate on why.

> Everything between "Yes it is a sort of equality test..." and "...I'm increasingly skeptical about is as an operator" looks like really far-fetched speculations to me, rooted rather in rhetorics than reality.

Unsurprisingly, I disagree. In my view, the survey provided some degree of statistical evidence regarding how people interpret things and the point about equality tests was brought up by many, and mistakes were made wrt. `is` and `as`.

I don't think such concerns should be so casually dismissed.

Ixrec commented:
FWIW, I'm not at all convinced that "pattern test expressions" (or whatever we want to call `let x: bool = (EXPR is PAT);`) are a feature that we want, much less that it should use the same syntax as `if let` chaining. In fact, I'm surprised that that's the thing dominating discussion so far, and the claim that this RFC is merely a "half-assed solution" sounds kind of extreme and bafflingly hostile to me. Fortunately, it doesn't sound like we actually need to settle that question to decide whether we want to do the edition change proposed here.

But more long term, if the survey described in the RFC and its hundreds of submissions are not considered sufficient evidence to "short-circuit" a bikeshed and just pick the syntax that has a clear majority behind it (unless someone has a reason to think the survey was not representative), _what would be enough_?

petrochenkov commented:
@Centril 
>The pattern is on the left and "short lived".

The pattern is on the right. Or in the end if you wish, so it's read last, regardless of formatting.
Using linguistic analogy, Latin script doesn't become right-to-left (or Arabic left-to-right) due to line wraps.

Centril commented:
@petrochenkov 

> The pattern is on the right. Or in the end if you wish, so it's read last, regardless of formatting.

I'm not sure about that; If I see:

```rust
if let Some(Def::Existential(_)) = tcx.describe_def(def_id)
    && let Some(node_id) = tcx.hir.as_local_node_id(def_id)
    && let hir::map::NodeItem(item) = tcx.hir.get(node_id)
    && let hir::ItemExistential(ref exist_ty) = item.node
    && let Some(parent) = exist_ty.impl_trait_fn
{
    ...
}
```

I will scan all `let` tokens in one go and notice them because of alignment. This entails that later `let`s are read before the expressions. Similarly, it is not necessarily the case that I scan `EXPR` in `match EXPR { .. }` before all the patterns in there.

>  Using linguistic analogy, Latin script doesn't become right-to-left (or Arabic left-to-right) due to line wraps.

🤔 My initial thought is that: Typography has some effects on how things are read which may include the order.

gilescope commented:
Could this not be done already as matching a tuple?

```
if let (Some(Def::Existential(_)), Some(node_id) = (tcx.describe_def(def_id), tcx.hir.as_local_node_id(def_id))
{
}
```

Ixrec commented:
@gilescope Yep, [that's addressed in the RFC](https://github.com/Centril/rfcs/blob/rfc/let-chains-2/text/0000-if-let-chains.md#other-solution-tuples):

> One solution is matching a tuple, but that is a poor solution when there are side effects or expensive computations involved, and doesn't necessarily work as DSTs and lvalues can't go in tuples.

vi commented:
Is `if !let Some(x) = ... { return Err(()) };  foo(x);`, which was also discussed somewhere before out of scope for this RFC?

Centril commented:
@vi For this RFC it is out of scope; I'd like to get the bare minimum accepted for now; we can revisit that later.

nikomatsakis commented:
So my overall feeling is that:

- I want to write this fairly regularly -- and `&&` is basically exactly the syntax I want, yes.
  - I recognize that more general alternatives exist (e.g., `is`, `matches`) -- but I feel like I rarely want them, and the keyword choice seems non-obvious. It seems like it would be ok to have `if let` with `&&` **and** such a syntax, just as it is ok to have `if let` and `match`.
- It certainly seems that `if let true = foo && bar` is a pretty unlikely pattern. It's tempting to think that we could change this without regard to edition and get away with it. But I also like doing things the principled way -- this is what editions are for, after all.
- On that point, I am a bit concerned about edition compatibility and macros here. I wish that we had our story around edition compatibility with macros fully straightened out so that we could specify it here. That said, I don't know if this concern needs to block this RFC per se -- it feels like a *somewhat* orthogonal concern.

So I think I feel vaguely positive about this RFC.

Centril commented:
@nikomatsakis I agree with all of your points; but I'm not precisely sure what macros are referring to in your comment? If you have an example, hopefully I could address it. :)

I60R commented:
What about `&&` support inside of `for` loops? E.g. the following snippet from RFC:

```rust
let mut results = [0, 0, 0, 0];
let inputs = [3, 7, 9, 6];

let r_iter = results.iter_mut();
let c_iter = 0..10;
let i_iter = inputs.iter();

while let Some(r) = r_iter.next()
    && let Some(index) = c_iter.next()
    && let Some(input) = i_iter.next()
{
    *r = index * 10 + input;
}
```

With for-chains it might be rewritten as:

```rust
let mut results = [0, 0, 0, 0];
let inputs = [3, 7, 9, 6];

for r in &mut results && index in 0..10 && input in inputs {
    *r = index * 10 + input;
}
```

kennytm commented:
@I60R 

1. that may be misinterpreted as a nested for loop
2. in reality you should just use `for (r, index, input) in izip!(&mut results, 0..10, inputs) {` or `for ((r, index), input) in results.iter_mut().zip(0..10).zip(inputs)` anyway
3. and this further complicates expression parsing

mark-i-m commented:
It also isn't at all clear what iterator that desugars to

Centril commented:
A further motivating example from https://github.com/rust-lang/rust/pull/52702:

```rust
if let TypeError::Mutability = terr
    && let Some(trait_err_span) = trait_err_span
    && let Ok(trait_err_str) = tcx.sess.codemap().span_to_snippet(trait_err_span)
{
    diag.span_suggestion(
        impl_err_span,
        "consider change the type to match the mutability in trait",
        format!("{}", trait_err_str),
    );
}
```

Also relevant: https://medium.com/programming-servo/programming-servo-how-to-match-b76c43f76fe6

joshtriplett commented:
I think we definitely want *some* way to chain `if let` conditions, and combine them with boolean conditions. This may or may not be the final syntax, but it seems compelling enough that we can proceed with this syntax in good faith unless we come up with some better alternative in the interim.

Given that:

@rfcbot fcp merge

nrc commented:
Adding my voice to the pile of 'this seems a good thing to do, but I'm not super-happy with the specifics'.

nox commented:
This RFC feels very unintuitive to me, things that bind names to values are very much not boolean expressions. One could say that Erlang's sugar for conjunctions and disjunctions in guard expressions is similar, and I've read *many* people complain about that along the years.

nox commented:
> In Erlang guards use ; for "OR" and , for "AND", while uses orelse and andalso in expressions. I don't see how the situation is similar to this RFC.

Yeah sorry I could have been clearer on that. I saw various samples in that discussion where people suggested being able to group `if let` bindings in parens like they were actual boolean expressions, with parentheses and whatnot. This happens often and causes frustration in Erlang where the guards aren't even using the same operators as in expressions, so I feel like the problem will be even worse in Rust land.

The fact that they get mixed with `&&` brings to the false intuition that things like `||` could work too, or arbitrary grouping and nesting, and frustration comes when user realises that's not possible.

Centril commented:
@nox 

> This RFC feels very unintuitive to me, things that bind names to values are very much not boolean expressions.

Pattern matching as a boolean expression is not unheard of and could be implemented in Rust if we so wished (using the currently proposed syntax or some other syntax...) with well defined semantics.
However, this RFC starts with a more conservative approach and attempts to solve the most pressing problem that many people have had over the years. And even with the conservative approach, I would say that the semantics of `if condition && let Some(x) = expr { .. }` should be fairly intuitive to deduce because one could imagine what it would mean if `let Some(x) = expr` was an expression.

> This happens often and causes frustration in Erlang where the guards aren't even using the same operators as in expressions, so I feel like the problem will be even worse in Rust land.

Let us try to measure and experiment :)
This RFC has already had a precursor as well as a survey that tested some of these hypotheses, and most people, *by far* felt that `&&` combined with `if` and `if let` was intuitive.

> The fact that they get mixed with `&&` brings to the false intuition that things like `||` could work too, or arbitrary grouping and nesting, and frustration comes when user realises that's not possible.

There is already frustration when the user notices that they can't combine `if let`s and `if`s using `&&` *today*. What we could do is to point users, who try to write `||` in these situations, to a tracking issue where they can explain their use cases. Then we can learn and improve the language from there.

nox commented:
> However, this RFC starts with a more conservative approach and attempts to solve the most pressing problem that many people have had over the years. And even with the conservative approach, I would say that the semantics of `if condition && let Some(x) = expr { .. }` should be fairly intuitive to deduce because one could imagine what it would mean `if let Some(x) = expr` was an expression.

This is unrelated to what I said. This RFC *does* make `&&` be used in other contexts than boolean expressions, for things that aren't boolean expressions. This is confusing to me.

> Let us try to measure and experiment :)

Let me state my experience with similar experiments or existing features in languages I've used in anger.

> This RFC has already had a precursor as well as a survey that tested some of these hypotheses, and most people, by far felt that `&&` combined with `if` and `if let` was intuitive.

The precursor was also heavily argued against by many people, and surveys sound like popularity contests to me.

> There is already frustration when the user notices that they can't combine `if let`s and `if`s using `&&` today. What we could do is to point users, who try to write `||` in these situations, to a tracking issue where they can explain their use cases.

But the rule is currently easy: `if let` takes a pattern and an expression and is *not* about boolean expressions. With this RFC, some constructions are special-cased (and change meaning compared to current Rust, like `if let true = foo && bar`), and some others are still not doable. The frustration will *increase*, IMO.

> Then we can learn and improve the language from there.

I don't consider reusing an existing syntax (joining boolean tests) for doing something else (binding values to names) an improvement, but that's just me.

comex commented:
> I don't consider reusing an existing syntax (joining boolean tests) for doing something else (binding values to names) an improvement, but that's just me.

Well, why not?

For context, I've been an advocate of making `let` a true boolean expression, but I'm happy with this RFC as an intermediate step.

Making `||` work is not only possible, it would be useful in some cases, e.g.:

```rust
if let Some(x) = get_x() ||
   let Some(y) = get_y() {
}   
```

And as has been discussed in other threads, it's also desirable to have a construct similar to "guard let" in Swift; I've personally had many use cases for this.  If `let` is a boolean, that can be `if !let`:

```rust
if !let Some(x) = get_x() {
    return;
}
// use x
```

So, that's `&&`, `||`, and `!`: three different operators.  We could define some new syntax for each of those, and/or leave some out.  But it seems more conceptually elegant to just make them booleans.  And I'm hoping that after people get experience using `&&` from this RFC, they'll want that as well. :)

Centril commented:
@nox You wrote:

> This is unrelated to what I said. This RFC does make `&&` be used in other contexts than boolean expressions, for things that aren't boolean expressions. This is confusing to me.

To which I replied:

> And even with the conservative approach, I would say that the semantics of `if` condition && let Some(x) = expr { .. }` should be fairly intuitive to deduce because one could imagine what it would mean `if let Some(x) = expr` was an expression.

That doesn't seem unrelated at all. I'm saying that even without `let PAT = EXPR` being a true boolean expression, it behaves as it would have had it been on in this context. Thus, the risk of confusion severely diminishes, in my opinion.

> Let me state my experience with similar experiments or existing features in languages I've used in anger.

Elaborate?

> The precursor was also heavily argued against by many people, and surveys sound like popularity contests to me.

Yes, it was, which is why this RFC proposes something else for good reasons, taking into account the many comments from the survey. I believe the survey was not such a contest because it received a high number of comments where people explained their thinking (some of which are included in the appendix and elsewhere in the text).

> But the rule is currently easy: `if let` takes a pattern and an expression and is not about boolean expressions. [..] The frustration will increase, IMO.

This RFC proposes a step in one direction and does so in a particular way because of time constraints.
It is not necessarily the end goal. We'll see where we end on that front.

I would also point to Sergio's comment at the top, https://github.com/rust-lang/rfcs/pull/2497#issuecomment-404705817, which notes that `||` is used with less frequency. I think that is right; wanting to use `&&` is more common, which explains why the number of people who have continuously argued for conjunctive chaining of `if let` are many more than those who have argued for disjunctive chaining.

> [..] (and change meaning compared to current Rust, like `if let true = foo && bar`), [..]

This bit has been covered in depth by this RFC. A survey of crates.io by Sergio found literally zero occurrences of this. Even if it had a few occurrences I think this is a bad thing to write and breaking it won't make me lose any sleep. Besides, we will issue a warning on Rust 2015 about it which rustfix can detect and fix for you by performing a trivial transformation (add parens on the RHS of the `=` sign).

> I don't consider reusing an existing syntax (joining boolean tests) for doing something else (binding values to names) an improvement, but that's just me.

In the end, I think we'll just have to agree to disagree on whether frustration will increase or not and whether this would be an improvement.

petrochenkov commented:
@comex 
>Making || work is not only possible, it would be useful in some cases, e.g.:

Except that `x` and `y` in the attached example are not usable because neither of them is "definitely initialized".
`||` with bindings is useful when those bindings are initialized on all paths through `||` (initialization dominates use).
```rust
if let Some(x) = get_x_in_one_way() ||
   let Some(x) = get_x_in_another_way() {
   use(x);
}  
```
(The primary issue is how to determine that both `x`s refer to the same variable and not separate variables.)

Centril commented:
> Except that `x` and `y` in the attached example are not usable because neither of them is "definitely initialized".

(I assume this was a typo on @comex's part, and it should have been:

```rust
if let Some(x) = get_x() ||
   let Some(x) = get_y() { // typo was in Some(y)
}
```
)

nox commented:
> ```rust
> if !let Some(x) = get_x() {
>     return;
> }
> // use x
> ```

My experience with this kind of feature is that it makes scoping rules quite confusing for users, this reminds me for example of variables exported from `case` expressions in Erlang. This is why I would rather have some sort of `guard` expressions tailored for that use case, rather than masquerading `let Pat = Expr` as some sort of fake boolean expression with complex scoping rules, which is what this eRFC proposes.

```erlang
case Status of
    1 ->
        Variable = "Something";
    2 ->
        Variable = "Something else";
    3 ->
        Variable = {"Something very different", [1,2,3]}
end
% Variable is now usable here. It would also work if Variable was defined in the pattern heads.
```

IMO, this eRFC makes the code especially hard to follow with `let Some(node_id) = tcx.hir.as_local_node_id(def_id) && let hir::map::NodeItem(item) = tcx.hir.get(node_id)`, where `node_id` is defined in a part of the test and used in the next one.

nox commented:
I think it would be better to have some sort of `with` construct a la Elixir, which would be easier to indent, easier to parse, and easier to explain given it doesn't change the meaning of existing constructs.

https://hexdocs.pm/elixir/Kernel.SpecialForms.html#with/1

nox commented:
At the very least, the `with` expressions from Elixir should be listed as an possible alternative and/or as prior art.

nox commented:
They are indeed similar, but I wouldn't expect anything starting with `if` to have weird scoping rules beyond what `if let` currently does (if it were me `if let` would never have been a thing and we would only have `with` but that's another story).

The "main alternatives" section is the result of a poll, and the construct you mention was "disliked". Dislike is about gut reactions, and we can't conclude that a `with` expression would have been as much disliked as piggy-backing the existing `if let` stuff.

Anyway, I somehow didn't see this alternative in the section you link, but I still think `with` expressions from Elixir should be mentioned in prior art.

shingtaklam1324 commented:
@kennytm

AFAIK Functionally there is no difference between what is proposed in the RFC apart from syntax/aesthetics. Assignment in Elixir is based on pattern matching (like Rust), so it runs through each statement in the with block, and only executes the statement in the do block if all of the pattern matching succeeds

avl commented:
I think irrefutable let bindings in the middle of a big if let statement could be very useful. Sometimes an expression becomes more readable if parts of it can be factored out into a named variable. You can't always bind such a variable before the if let, since it may depend on earlier bound variables in the same if let.

Centril commented:
**Huzzah!** This RFC has been **merged!**

Tracking issues:
+ Main issue: https://github.com/rust-lang/rust/issues/53667
+ Edition transition issue: https://github.com/rust-lang/rust/issues/53668

H2CO3 commented:
@Centril I somehow couldn't find in this RFC if the sudden precedence change of the `&&` and `||` operators *always* happens in the control expression of `if` and `while`, or only when there's at least one `let` binding? And if there's any more changes to operator precedence which have unintended consequences? The text of the rendered proposal says:

> To be precise, the changes in this RFC entail that `||` has the lowest precedence at the top level of `if STUFF { .. }`. The operator `&&` has then the 2nd lowest precedence and binds more tightly than `||`. If the user wants to disambiguate, they can write `(EXPR && EXPR)` or `{ EXPR && EXPR }` explicitly.

The phrasing here is not particularly clear (e.g. what is `STUFF`? what exactly does the user want to disambiguate?). To me it kind of suggests that the precedence change *always* happens and that from now on, parentheses will be basically obligatory around the top-level (non-`let`) expression. I'm also confused because I don't understand why the latter change is necessary (and if it is, I would find it very annoying).

H2CO3 commented:
@Centril Thanks for the explanation!

Centril commented:
@H2CO3 Welcome :)

phaazon commented:
Argh, I didn’t see that RFC and I wish I did… I’m not a huge fan of the proposal, it seems too niche and introduces too much sugar (mixes two very different constructs — `if _`, and « `if_let _` », blending `_: bool` with `_: ???` ). But the RFC is merged, so ignore this comment I guess. :disappointed:

Centril commented:
This is definitely not niche; in particular, this pattern arises frequently in the compiler (similar code like proc macros, clippy, anything that deals with AST-looking trees) and will pay for itself many times over. `if` and `if let` are not very different constructs; in particular, when `let pat = expr` is considered as an expression itself, then the distinction between these constructs evaporate completely making the language conceptually simpler in the process.

phaazon commented:
Yes, but `let pat = expr` can be considered an expression ONLY when it implies destructuring, otherwise it doesn’t make any sense (as it’s a statetement). With what you say, I would expect this to be valid:

```
let x = Some("foo");
let y = true || let Some(foo) = x;
```

Which is definitely not part of the language (and shouldn’t, as introducing bindings in expressions seems completely out of question).

So, yeah, maybe you’re introducing “conceptually simpler” sugar for very specific cases (again, I truly think it’s niche, and you put on a name on that niche: compilers and ASTs), but you’re **also** introducing confusion pretty much everywhere else where that “conceptually simpler” concept would feel applicable (while it’s not).

phaazon commented:
I would also like to mention that `if let pat = expr { _1 } else { _2 }` can be seen as a shortcut to `match pat { expr => { _1 }, _ => _1`. So to me, it’s not very the same construct as a single `if`, which is a simple instruction that is pretty simple to understand.

Nemo157 commented:
Just yesterday I wanted this feature while [changing the MPSC channels in `futures`](https://github.com/rust-lang-nursery/futures-rs/pull/1443/files#diff-8fe465859d06839f1c81c2a1ae56c0fcR710), that's very much not the compilers/ASTs niche. If you really want I'm sure I can find some more examples from other non-compiler/AST projects I have worked on.

phaazon commented:
In the code you showed, you could have used a guarded match. That’s already part of the language and it will do the job for most cases. Having lazy pattern-matching mixed with booleans makes me incomfortable. I’m not saying I’m against lazy pattern-matching (it’s actually handy). I’m just saying that using the `if` keyword for that is not a good move. But as I said yesterday, the RFC is already merged and we’re just bikeshedding. _Tant pis pour moi_

kennytm commented:
@alper It was considered and the community highly disliked it. There is literal [a survey report](https://github.com/rust-lang/rfcs/blob/master/text/2497-if-let-chains.md#a-survey---data) inside the RFC document.
